pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/invertedindex/RealtimeLuceneIndexReaderRefreshThread.java

```diff
@@ -0,0 +1,133 @@
+ * readers as the number of columns with text search enabled.
+ *
+ * Between each successive execution of the task, there is a fixed delay (regardless of how long
+ * each execution took). When the task wakes up, it pick the RealtimeLuceneReadersForRealtimeSegment
+ * from the head of queue, refresh it's readers and adds this at the tail of queue.
+ */
+public class RealtimeLuceneIndexReaderRefreshThread implements Runnable {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentColumnarIndexCreator.class);
+  // TODO: make this configurable and choose a higher default value
+  public static final int DELAY_BETWEEN_SUCCESSIVE_EXECUTION_MS_DEFAULT = 10;
+
+  private final ConcurrentLinkedQueue<RealtimeLuceneReadersForRealtimeSegment> _luceneRealtimeReaders;
+  private final Lock _mutex;
+  private final Condition _conditionVariable;
+
+  private volatile boolean _stopped = false;
+
+  RealtimeLuceneIndexReaderRefreshThread(
+      ConcurrentLinkedQueue<RealtimeLuceneReadersForRealtimeSegment> luceneRealtimeReaders,
+      Lock mutex, Condition conditionVariable) {
+    _luceneRealtimeReaders = luceneRealtimeReaders;
+    _mutex = mutex;
+    _conditionVariable = conditionVariable;
+  }
+
+  void setStopped() {
+    _stopped = true;
+  }
+
+  @Override
+  public void run() {
+    while (_stopped) {
+      _mutex.lock();
+      while (_luceneRealtimeReaders.isEmpty()) {
+        try {
+          // During instantiation of MutableSegmentImpl, we will signal on this condition variable once
+          // one or more realtime lucene readers (one per column) belonging to the MutableSegment
+          // are added to the global queue managed by this thread. The thread that signals will
+          // grab this mutex and signal on the condition variable.
+          //
+          // This refresh thread will be woken up (and grab the mutex automatically as per the
+          // implementation of await) and check if the queue is non-empty. It will then proceed to
+          // poll the queue and refresh the realtime index readers for the polled segment.
+          //
+          // The mutex and condition-variable semantics take care of the scenario when on
+          // a given Pinot server, there is no realtime segment with text index enabled. In such
+          // cases, there is no need for this thread to wake up simply after every few seconds/minutes
+          // only to find that there is nothing to be refreshed. The thread should simply be
+          // off CPU until signalled specifically. This also covers the situation where initially
+          // there were few realtime segments of a table with text index. Later if they got
+          // moved to another server as part of rebalance, then again there is no need for this thread
+          // to do anything until some realtime segment is created with text index enabled.
+          _conditionVariable.await();
```

Comment:
we should check for `_stopped` after exiting from the condition variable and break out

Reply:
I think we should also check for the q not being empty upon exiting the cond var to handle spurious wakeups, then unlock and loop back.

Reply:
Checking for _stopped.

The spurious wakeup is already addressed since we wait on the condition variable in a while loop